File: 剑指offer/24.反转链表.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        new_head = ListNode(0)
        p = head

        while p:
            next = p.next
            p.next = new_head.next
            new_head.next = p

            p = next

        return new_head.next

File: 剑指offer/24.反转链表.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};

class Solution {
public:
    ListNode* reverseList(ListNode* head)
    {
        ListNode* new_head = new ListNode();

        ListNode* p = head;

        while (p != NULL) {
            ListNode* next = p->next;

            p->next = new_head->next;
            new_head->next = p;

            p = next;
        }

        return new_head->next;
    }
};